documentation/specs/proposed/BuildCheck-Architecture.md

```diff
@@ -0,0 +1,88 @@
+
+The BuildCheck infrastructure will be prepared to be available concurrently within the `entrypoint node` as well as in the additional `worker nodes`. There are 2 reasons for this:
+* BuildCheck will need to recognize custom analyzers packages during the evaluation time - so some basic code related to BuildCheck will need to be present in the worker node.
+* Presence in worker node (as part of the `RequestBuilder`), will allow inbox analyzers to agile leverage data not available within `BuildEventArgs` (while data prooved to be useful should over time be exposed to `BuildEventArgs`)
+
+## Handling the Distributed Model
+
+We want to get some benefits (mostly inbox analyzers agility) from hosting BuildCheck infrastructure in worker nodes, but foremost we should prevent leaking the details of this model into public API and OM, until we are sure we cannot achive all goals from just entrypoint node from `BuildEventArgs` (which will likely never happen - as the build should be fully reconstructable from the `BuildEventArgs`).
+
+How we'll internally handle the distributed model:
+* Each node will have just a single instance of infrastructure (`IBuildCheckManager`) available (registered via the MSBuild DI - `IBuildComponentHost`). This applies to a entrypoint node with inproc worker node as well.
+* Entrypoint node will have an MSBuild `ILogger` registered that will enable communicating information from worker nodes BuildCheck module to the entrypoint node BuildCheck module - namely:
+    * Acquisition module from worker node will be able to communicated to the entrypoint node that it encountered `PackageReference` for particular analyzer and that it should be loaded and instantiated in the main node.
+    * Tracing module will be able to send perf stats from current worker node and aggregate al of those together in the main node.
+    * Theoretical execution-data-only sourcing inbox analyzer will be able to aggregate data from the whole build context (again - we should use this only for agility purposes, but shoot for analyzer that needs presence only in entrypoint node). The way to do that can be via being present in all worker nodes, sending a specific type of 'in progress result' BuildEventArgs and aggreggating those intermediary results in the single instance running in the main node.
+* Apart from the scenarios above - the BuildCheck infrastructure modules in individual nodes should be able to function independently (namely - load the inbox analyzers that should live in nodes; send the analyzers reports via logging infrastructure; load user configuration from `.editorconfig` and decide on need to enable/disable/configure particular analyzers).
+* Communication from main to worker node between BuildCheck infra modules is not planned.
+
+## Analyzers Lifecycle
+
+Planned model:
+* Analyzers factories get registered with the BuildCheck infrastructure (`BuildCheckManager`)
```

Comment:
Is there a better name than `BuildCheckManager`? I feel like class names ending with `*Manager` can generally be given more descriptive names.

What about BuildCheck does it manage?

Reply:
It serves as the pluged component into the build that receives events and data from the build and based on those de/registers, instantiates, re/configures, runs the actual analyzer instances. So `BuildCheckEngine` or `BuildCheckInfrastructure` were another ideas - but that feels equaly broad.

Other option is no suffix - as `BuildCheck` was originaly meant as the codename for the module within the MSBuild that hosts `BuildAnalyzer`s. But if we are inclining to use the term `BuildCheck` for the later term - this would be out of question.